### PR TITLE
shell: Purge iframes for inaccessible machines

### PR DESCRIPTION
**Pull request by petervo**
This is part of the work to make handling reboots smoother.

When a connection to a machine is lost navigate away from it and purge it's iframes. Navigation back to that machine will attempt to reestablish a connection.

When attempting to access a known but disconnected machine i opted to begin loading while it attempts to get the connection rather that sitting and waiting, we may want something designed for that state.

Another question i had was do we want to use patternfly alerts to notify users about the disconnection for connections that are not monitored by the watchdog.

### CODE REVIEW

**Comment by andreasn** [comment]
"When attempting to access a known but disconnected machine i opted to begin loading while it attempts to get the connection rather that sitting and waiting, we may want something designed for that state."

Maybe a spinner would be best for that.

**Comment by andreasn** [comment]
Some designs for indications on what's going on.
Mockup

**Comment by petervo** [comment]
Added the rest of the work for the unexpected disconnection case. I didn't yet add the dashboard spinners and message for expected restarted.

@stefwalter to make this as well as the coming expected restarts work some communication is needed between the frames. I did the first part of it here without making api changes to cockpit and cockpit transport. So it's fairly specific to this use case. Before I do the expected restarts do you want to take a look and see if you think this is the right approach for now, or if i should add something more permanent to cockpit.js.

**Comment by stefwalter** [comment]
> @stefwalter to make this as well as the coming expected restarts work some communication is needed between the frames. I did the first part of it here without making api changes to cockpit and cockpit transport. So it's fairly specific to this use case. Before I do the expected restarts do you want to take a look and see if you think this is the right approach for now, or if i should add something more permanent to cockpit.js.

Having additional communication makes sense. Maybe a 'restarting' message to the index frame would make sense? See the #1563 for sharing data between index.js and cockpit-dashboard.js

**Comment by petervo** [comment]
Updated to be based on #1563

**Comment by petervo** [comment]
Rebased 

@andreasn added ui for expected restarts

**Comment by petervo** [comment]
depends on #2400

**Comment by andreasn** [comment]
Looks good!

**Comment by petervo** [comment]
Not sure how this got closed

**Comment by petervo** [comment]
@stefwalter The last 2 commits are more about fixing some issues with the json-cache machine implementation. If this isn't going to make it in soon we maybe should split those two off into a separate PR.

File: pkg/base1/cockpit.js [hunk 1]
```diff
@@ -926,6 +926,29 @@ function basic_scope(cockpit) {
...
+
+        var timer = null;
+        var options = { "command": "expect-restart",
+                        "host": host };
+
+        function send() {
+            ensure_transport(function(transport) {
+                transport.send_control(options);
+                if (timer)
+                    window.clearTimeout(timer);
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: What is happening here? Why clearTimeout? Hasn't the timer already fired at this point?

File: pkg/base1/cockpit.js [hunk 2]
```diff
@@ -926,6 +926,29 @@ function basic_scope(cockpit) {
         });
     };
 
+    /* Not public API ... yet? */
+    cockpit.expect_restart = function expect_restart(delay, host) {
```
Comment (stefwalter) [inline pkg/base1/cockpit.js]: Do you think that we have an expect API? For various global events? And then have 'restart' as a code?

File: pkg/shell/machines.js [hunk 3]
```diff
@@ -161,10 +184,30 @@ define([
             if (machine) {
                 delete machines[host];
                 self.disconnect(host);
-                notify(host, "removed");
+                notify(host);
             }
         }
 
+        self.expect_restart = function expect_restart(host) {
```
Comment (stefwalter) [inline pkg/shell/machines.js]: Even if we do have a more general ```cockpit.expect``` API (noted above) ... then having this as a specific API is still good.

File: pkg/shell/machines.js [hunk 4]
```diff
@@ -161,10 +184,30 @@ define([
...
+            function reconnect () {
+                machine.restarting = true;
+                var timer = window.setTimeout(function() {
+                    window.clearTimeout(timer);
```
Comment (stefwalter) [inline pkg/shell/machines.js]: Why are we calling clearTimeout in a timeout callback?

File: pkg/shell/cockpit-server.js [hunk 5]
```diff
@@ -868,6 +868,13 @@ PageShutdownDialog.prototype = {
                 shell.show_unexpected_error(ex);
             })
             .done(function(ex) {
+                var expect_delay = parseInt(delay);
```
Comment (stefwalter) [inline pkg/shell/cockpit-server.js]: Hmmm, not sure about the whole concept of delays. These could be days away. If this is the only use case of the concept of delays in expect restart, then I would skip on it for now.

For this particular case we need to listen to the logind Dbus API and get our restart information from there. Once we're at around 60 seconds, just tell the outer frame that a restart is expected and leave it at that, no?

Again, such work should be a follow up. I'm just leary of cementing the concept of delays into the API.

File: pkg/shell/index.html [hunk 6]
```diff
@@ -186,6 +186,12 @@ <h4 class="panel-title">
       </div>
     </div>
 
+    <div id="connecting" style="display: none;">
+        <div class="spinner spinner-lg"></div>
+        <br/>
+        <p>Trying to reconnect to the system</p>
```
Comment (stefwalter) [inline pkg/shell/index.html]: Needs to be translatable.

File: pkg/shell/index.js [hunk 7]
```diff
@@ -205,6 +205,9 @@ define([
     var router = new Router();
     var packages = new Packages();
 
+    var checksums = {};
+    var machine_components = {};
```
Comment (stefwalter) [inline pkg/shell/index.js]: I would suggest making machine_components private to the Components object.

File: pkg/shell/index.js [hunk 8]
```diff
@@ -217,20 +220,34 @@ define([
 
     /* When the machine list is ready we start processing navigation */
     $(machines)
+       .on("read-complete", function(ev) {
```
Comment (stefwalter) [inline pkg/shell/index.js]: How is "read-complete" different from "ready"?

File: pkg/shell/index.js [hunk 9]
```diff
@@ -336,14 +362,20 @@ define([
...
+            machine = machines.lookup(address);
 
             /* If the machine is not available, then redirect to dashboard */
-            machine = machines.lookup(address);
-            if (!machine) {
-                if (listing)
+            if (!machine || machine.problem) {
+                if (!machine && !machines.read_complete) {
+                    show_connecting();
+                } else if (machine && reconnect && machine.problem) {
```
Comment (stefwalter) [inline pkg/shell/index.js]: Shouldn't we track this off of the machines object and not expect the value to be passed into navigate()?

File: pkg/shell/index.js [hunk 10]
```diff
@@ -711,6 +760,12 @@ define([
                         perform_jump(control);
                     return;
 
+                } else if (control.command === "expect-restart") {
+                    /* watchdog handles current host for now */
+                    if (control.host != cockpit.transport.host)
```
Comment (stefwalter) [inline pkg/shell/index.js]: Shouldn't a null here mean that the default_host of the component which sent the message is the one that's restarting?

File: pkg/shell/index.js (hunk 8 above)
Comment (petervo) [inline pkg/shell/index.js]: Ready gets fired the moment the 1st message comes in. Read complete is when processing triggered by a file change completes.

File: pkg/shell/index.js (hunk 10 above)
Comment (petervo) [inline pkg/shell/index.js]: cockpit.hint fills in the host, it's not going to be null here.

File: pkg/shell/index.js (hunk 9 above)
Comment (petervo) [inline pkg/shell/index.js]: It's not really about the machine but about if the navigation call is triggered by an action or triggered by one of the machine event callbacks. Machine is shared around, and this information is very local to what index is doing with navigation.

**Comment by petervo** [comment]
Made most of the changes except for the 2 I commented on.

**Comment by stefwalter** [comment]
This will need a lot of further work once we have diagnosis based adding of machines to the dashboard (and kube nodes). But for now this is in the right direction.

**Pull request closed without merging**